subsys/net/l2/ppp/ppp_l2.c

```diff
@@ -284,11 +284,15 @@ static void carrier_on_off(struct k_work *work)
 		ppp_mgmt_raise_carrier_on_event(ctx->iface);
 		net_if_up(ctx->iface);
 	} else {
-		ppp_lower_down(ctx);
-		ppp_change_phase(ctx, PPP_DEAD);
+		if (ppp_lcp) {
```

Comment:
If I change this part to:

```c
		if (ppp_lcp) {
			ppp_lcp->close(ctx, "Shutdown");
			/* signaling for the carrier off event is done from the LCP callback */
		}
		ppp_change_phase(ctx, PPP_DEAD);

		ppp_mgmt_raise_carrier_off_event(ctx->iface);
		net_if_carrier_down(ctx->iface);
```

then the modem is able to start during the second cycle.

I added `NET_WARN("lcp_finished");` to `lcp_finished` but the line is not printed, maybe `lcp_finished` didn't run for some reason?

Reply:
Thanks for looking into it. The challenge is that I have a different implementation of the layers on higher level (proprietary legacy). This makes it slightly different in behavior.

The problem you are dealing with it that you shutdown the modem too early. So it did not have time to finish the sequence. This means that your GSM MUX is not switched properly back to AT mode and it will never respond to AT mode.
In my shutdown sequence I call the ppps_close function, then wait for a semaphore. This semaphore is being raised when the NET_EVENT_IF_DOWN event comes from the net_mgmt_event_callback. This is only triggered by the net_if_carrier_down(ctx->iface); being called in the lcp_finished function. This way you allow the state machine to properly finish the terminate sequence before cutting the power. This will even be reached when the number of retries is exhausted.

With respect to your last suggestion. This will cancel the effect of the new changes will be cancelled. The ppp_lcp->close will put the packet in the TX queue, but by the time the queue is being severed after this function, the interface has been shut down, so no packet will be send and you will probably bump into the line: NET_WARN("iface %p is down", iface); in net_if.c

It obviously needs some more work to work with the stock example. Due to COVID I have no access to HW and debugger today. If you could point me to an example from the examples folder I can try to make it work on my HW.

Reply:
Indeed I turned off my modem too quickly, as sometimes it doesn't reply to the shutdown command/restart itself after shut down.

I don't really have an example, but a simple way to test it:

pseudo code:
```c
while (true) {
        enable_modem_power();
        gsm_ppp_start();
        k_msleep(60000);
        gsm_ppp_stop();
        disable_modem_power();
        k_msleep(20000);
}

Reply:
Just looking at the gsm_ppp_stop function, I think to make it work properly a semaphore construction should be between, to wait for the interface to be properly closed off (NET_EVENT_IF_DOWN) event:

`net_if_l2(iface)->enable(iface, false);`

and 

```
if (IS_ENABLED(CONFIG_GSM_MUX)) {
		/* Lower mux_enabled flag to trigger re-sending AT+CMUX etc */
		gsm->mux_enabled = false;

		if (gsm->ppp_dev) {
			uart_mux_disable(gsm->ppp_dev);
		}
	}
```

The ->enable(iface, false) triggers the close, but the gsm->mux_enabled = false; needs to be delayed for the FSM to finish its termination.

This construction introduces a dependency between the GSM_PPP driver and the NET_MGMT_EVENT. But on the other hand it enables someone to use the GSM_PPP driver out of the box.
An alternative is to poll for the interface to go down.

Reply:
> The ->enable(iface, false) triggers the close, but the gsm->mux_enabled = false; needs to be delayed for the FSM to finish its termination.

Indeed, good catch. @ssharks would you be able to send a fix for this?

Reply:
I can give it a try. It will be earliest next week as I need some HW to test this. Next the GSM_PPP driver is quite device specific at the moment, so I need to adjust it to the device I have as well.

Do you want to have it in a separate commit, i.e. separate pull request or how to send in these separate but somewhat related changes.